lib/statsd/instrument/udp_sink.rb

```diff
@@ -0,0 +1,45 @@
+# frozen_string_literal: true
+
+class StatsD::Instrument::UDPSink
+  def initialize(host, port)
+    @host = host
+    @port = port
+    @mutex = Mutex.new
+    @socket = nil
+  end
+
+  def <<(datagram)
+    with_socket { |socket| socket.send(datagram, 0) > 0 }
+    self
+
+  rescue ThreadError
+    # In cases where a TERM or KILL signal has been sent, and we send stats as
+    # part of a signal handler, locks cannot be acquired, so we do our best
+    # to try and send the datagram without a lock.
+    socket.send(datagram, 0) > 0
+
+  rescue SocketError, IOError, SystemCallError
+    # TODO: log?
+    invalidate_socket
+  end
+
+  private
+
+  def with_socket
+    @mutex.synchronize { yield(socket) }
```

Comment:
Kind of note to self: it would be interesting to measure the cost of that mutex, and compare with potential alternatives (e.g. per thread socket).

Reply:
Perhaps we should do the synchronization outside of the sync so that we can easily choose to have client per-thread.  We could have a SynchronizedSync that wraps the `UDPSink` with this mutex synchronization.